src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs

```diff
@@ -324,30 +326,33 @@ private static IResourceBuilder<ProjectResource> WithProjectDefaults(this IResou
         }
         else
         {
-            // If we aren't a web project (looking at both launch profile and Kestrel config) we don't automatically add bindings.
-            if (launchProfile?.ApplicationUrl == null && !kestrelEndpointsByScheme.Any())
+            // Set HTTP_PORTS/HTTPS_PORTS in publish mode, to override the default port set in the base image. Note that:
+            // - We don't set them if we have Kestrel endpoints configured, as Kestrel will get everything from its config.
+            // - We only do that for endpoint set explicitely (.WithHttpEndpoint), not for the ones coming from launch profile.
+            //   This is because launch profile endpoints are not meant to be used in production.
+            if (!kestrelEndpointsByScheme.Any())
             {
-                return builder;
+                builder.SetBothPortsEnvVariables();
             }
 
-            if (!projectResource.Annotations.OfType<EndpointAnnotation>().Any(sb => sb.UriScheme == "http" || string.Equals(sb.Name, "http", StringComparisons.EndpointAnnotationName)))
+            // If we aren't a web project (looking at both launch profile and Kestrel config) we don't automatically add bindings.
+            if (launchProfile?.ApplicationUrl == null && !kestrelEndpointsByScheme.Any())
             {
-                builder.WithEndpoint("http", e =>
-                {
-                    e.UriScheme = "http";
-                    e.Transport = adjustTransport(e);
-                },
-                createIfNotExists: true);
+                return builder;
             }
 
-            if (!projectResource.Annotations.OfType<EndpointAnnotation>().Any(sb => sb.UriScheme == "https" || string.Equals(sb.Name, "https", StringComparisons.EndpointAnnotationName)))
+            string[] schemes = ["http", "https"];
```

Comment:
This part is just a little refactoring to avoid repeating code, and not directly related to this issue